packages/logger/src/types/Log.ts

```diff
@@ -3,10 +3,10 @@ type LogLevelInfo = 'INFO';
 type LogLevelWarn = 'WARN';
 type LogLevelError = 'ERROR';
 
-type LogLevel = LogLevelDebug | LogLevelInfo | LogLevelWarn | LogLevelError | string;
+type LogLevel = LogLevelDebug | Lowercase<LogLevelDebug> | LogLevelInfo | Lowercase<LogLevelInfo> | LogLevelWarn | Lowercase<LogLevelWarn> | LogLevelError | Lowercase<LogLevelError>;
```

Comment:
Is it possible to extend the types with `type LogLevelDebug = 'DEBUG' | 'debug';` to simplify it? With that we only need two `toUpperCase()` conversions: in `setLogger` and `shouldPrint`. 

Curious to understand why `Uppercase` and `Lowercase` generics were introduced.

Reply:
I initially started out the same way you suggested. However later on I realized this would clash with how the levels are actually used internally in the Logger. 

The [logic that decides](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/feat/logger/stricter_loglevel_types/packages/logger/src/Logger.ts#L857) on whether a log should be emitted uses numeric values to do so, and these numeric values are [defined here](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/feat/logger/stricter_loglevel_types/packages/logger/src/Logger.ts#L133-L138).

So essentially, to avoid changing the logic too much and have to deal internally with both uppercase & lowercase, I introduced these generics so that Logger is able to accept both upper & lowercase but internally it's all still treated as [if it was uppercase only](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/feat/logger/stricter_loglevel_types/packages/logger/src/types/Log.ts#L8-L10).

Reply:
The `LogLevelThresholds` with `shouldPrint` is the only place were we need to convert to uppercase. Using `setLogger` we can enforce the uppercase type value. Types only help us to handle user input when passing value to Logger constructor. After we created the object, there is no need to convert them during the runtime. Here is an example of `setLogLevel`

```ts
if (this.isValidLogLevel(logLevel)) {
      this.logLevel = logLevel?.toUpperCase() as LogLevel;

      return;
    }
```
After we know that we got a valid `LogLevel` we set it to uppercase. With this we can have

```ts
if (
      this.logLevelThresholds[logLevel.toUpperCase() as Uppercase<LogLevel>] >=
      this.logLevelThresholds[this.getLogLevel().toUpperCase() as Uppercase<LogLevel>]
    ) {
      return true;
    }
``` 

Technically the `logLevel.toUpperCase() as Uppercase<LogLevel>` is unnecessary, because we know it will be uppercase, we set it during `setLogLevel`.

`LogLevelThresholds` type remains:
```ts
type LogLevelThresholds = {
  [key in Uppercase<LogLevel>]: number;
};
```

With this we can remove some type assertions of `Uppercase` and `Lowercase`, but not all of them.

I think my main concern is that Lowercase and Uppercase assertions feels scattered across the codebase. In `Log` type we have lowercase, and in `Logger` it's all uppercase, and then we call `toUppercase()` in many places, and I'd like to avoid them.

Reply:
I see what you mean and I agree.

During the refactor I mistakenly thought that the instance of `logLevel` was user provided and that as such we should enforce its case. However as you've pointed out this is not the case and we already know it's uppercase.

In [this commit](https://github.com/awslabs/aws-lambda-powertools-typescript/pull/1313/commits/2ed61c677a0c0121f60f890326256aaea3176193) I have fixed the input type of that function and removed the typecast. 

Now the only place where `Lowercase` is used is in the `Log` types, whereas elsewhere in the `Logger` class we use always `Uppercase`. Based on my current understanding of these generics, since `LogLevel` is now mixed, we need to use `Uppercase` to tell TS that we know/want it's uppercase.